src/MmappedFileMonitor.cc

```diff
@@ -0,0 +1,67 @@
+/* -*- Mode: C++; tab-width: 8; c-basic-offset: 2; indent-tabs-mode: nil; -*- */
+
+#include "MmappedFileMonitor.h"
+
+#include "RecordSession.h"
+#include "RecordTask.h"
+#include "log.h"
+
+using namespace std;
+
+namespace rr {
+
+MmappedFileMonitor::MmappedFileMonitor(Task* t, int fd) {
+  ASSERT(t, !t->session().is_replaying());
+  stat_ = t->stat_fd(fd);
+}
+
+MmappedFileMonitor::MmappedFileMonitor(Task* t, EmuFile::shr_ptr f) {
+  ASSERT(t, t->session().is_replaying());
+  emu_file_ = f;
+}
+
+void MmappedFileMonitor::did_write(Task* t, const std::vector<Range>& ranges,
+				   int64_t offset) {
+  if (ranges.empty()) {
+    return;
+  }
+
+  ASSERT(t, offset >= 0)
+      << "Only pwrite/pwritev supported on /proc/<pid>/mem currently";
+  // XXX to fix that, we'd have to grab the file offset from
+  // /proc/<pid>/fdinfo.
+
+  bool is_replay = t->session().is_replaying();
+  for (auto m : t->vm()->maps()) {
+    auto km = m.map;
+
+    if (is_replay) {
+      if (!emu_file_ || emu_file_ != m.emu_file) {
+        continue;
+      }
+    } else {
+      if (km.device() != stat_.st_dev || km.inode() != stat_.st_ino) {
+        continue;
+      }
+    }
+
+    // stat matches.
+    uint64_t mapping_offset = km.file_offset_bytes();
+    int64_t local_offset = offset;
+    for (auto r : ranges) {
+      remote_ptr<void> start = km.start() + local_offset - mapping_offset;
+      MemoryRange mr(start, r.length);
+      if (km.intersects(mr)) {
+        if (is_replay) {
+          // If we're writing beyond the EmuFile's end, resize it.
+          emu_file_->ensure_size(local_offset + r.length);
+        } else {
+          static_cast<RecordTask*>(t)->record_remote(km.intersect(mr));
```

Comment:
If the file is mapped into multiple places, this will record multiple writes. That's not great but I guess we can live with it for now. Just mention it in a comment.